x/nft/username/handler.go

```diff
@@ -249,3 +249,71 @@ func findActor(h tokenHandler, ctx weave.Context, t Token) weave.Address {
 	}
 	return nil
 }
+
+type AddRemoveApprovalHandler struct {
+	tokenHandler
+}
+
+func (h *AddRemoveApprovalHandler) Check(ctx weave.Context, store weave.KVStore, tx weave.Tx) (weave.CheckResult, error) {
+	var res weave.CheckResult
+	if _, err := h.validate(ctx, tx); err != nil {
+		return res, err
+	}
+	res.GasAllocated += createUsernameCost
+	return res, nil
+}
+
+func (h *AddRemoveApprovalHandler) Deliver(ctx weave.Context, store weave.KVStore, tx weave.Tx) (weave.DeliverResult, error) {
+	var res weave.DeliverResult
+	msg, err := h.validate(ctx, tx)
+	if err != nil {
+		return res, err
+	}
+
+	o, t, err := loadToken(h.tokenHandler, store, msg.GetId())
+	if err != nil {
+		return res, err
+	}
+
+	actor := findActor(h.tokenHandler, ctx, t)
+	if actor == nil {
+		return res, errors.ErrUnauthorized()
+	}
+
+	switch v := interface{}(msg).(type) {
+	case *AddApprovalMsg:
+		base := v.Base
+
+		err = t.Approvals().Grant(base.Action, base.Address, base.Options, 0)
+		if err != nil {
+			return res, err
+		}
+	case *RemoveApprovalMsg:
+		base := v.Base
+
+		err = t.Approvals().Revoke(base.Action, base.Address)
+		if err != nil {
+			return res, err
+		}
+	}
+
+	return res, h.bucket.Save(store, o)
+}
+
+func (h *AddRemoveApprovalHandler) validate(ctx weave.Context, tx weave.Tx) (nft.Identified, error) {
```

Comment:
pp: the Identified interface is used only here. therefore I would move it into this file or just return the ID from this method

Reply:
@alpe not true, this interface is used elsewhere in your code, I just didn't refactor it.
Not sure if it'd be needed with the new approach

Reply:
done, I refactored your code to use it too

Reply:
Also, we will probably use it for anything that has GetId() in nft package anyway as needed